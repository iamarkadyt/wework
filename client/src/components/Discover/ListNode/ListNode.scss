.ListNode-container {
    margin-top: 1rem;
    display: grid;
    grid-template-columns: 3rem auto 1.4rem;
    grid-column-gap: .5rem;
    align-items: center;
    .ListNode-img {
        width: 3rem;
        border-radius: 3rem;
        height: 3rem;
    }
    .ListNode-name {
        p {
            &:first-child {
                font-size: 1.2rem;
                font-weight: bold;
            }
        }
    }
}
