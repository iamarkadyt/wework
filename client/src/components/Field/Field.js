import React from 'react'
import DropdownList from 'react-widgets/lib/DropdownList'
import DateTimePicker from 'react-widgets/lib/DateTimePicker'
import Multiselect from 'react-widgets/lib/Multiselect'
import Moment from 'moment'
import momentLocalizer from 'react-widgets-moment'
import './Field.scss'
import { node, string, func, object, bool, number, arrayOf, oneOfType, instanceOf } from 'prop-types'

Moment.locale('en')
momentLocalizer()

const field = ({
    // different fields require different props
    type,
    name, value, onChange, // all except submit and button
    placeholder, error, // all except checkbox, submit and button
    label, // checkbox, submit, button
    list, // list, multiselect
    rows, // textarea
    disabled, // date
    onClick, style, // button, linkButton, textarea
    inline, // container
    containerStyle, // container
    children
}) => {
    let field = null
    const commonProps = { name, id: name, value, onChange }

    switch (type) {
        case "textarea":
            field = <textarea
                {...commonProps}
                placeholder={placeholder || 'Say something...'}
                className="rw-widget-container rw-widget-input"
                style={{
                    padding: '.5rem',
                    fontSize: '1rem',
                    width: '100%',
                    resize: 'none',
                    ...style
                }}
                rows={rows} />
            break
        case "date":
            field = <DateTimePicker
                {...commonProps}
                placeholder={placeholder || 'Pick a date...'}
                time={false}
                format="MMM DD, YYYY"
                editFormat="M/DD/YYYY"
                disabled={disabled} />
            break
        case "list":
            field = <DropdownList
                {...commonProps}
                placeholder={placeholder || "Choose one..."}
                data={list} />
            break
        case "multiselect":
            field = <Multiselect
                {...commonProps}
                placeholder={placeholder || "Enter a few items..."}
                data={list} />
            break
        case "checkbox":
            field = <label htmlFor={name}>
                <input type="checkbox" {...commonProps} />
                &nbsp;&nbsp;{label}
            </label>
            break
        case "submit":
            field = <button
                type="submit"
                className="button-action">
                {label || children}
            </button>
            break
        case "button":
            field = <button
                className="button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        case "linkButton":
            field = <button
                className="link-button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        default:
            field = <input
                {...commonProps}
                placeholder={placeholder}
                className="rw-widget-container rw-widget-input"
                type={type} />
    }

    return (
        <div style={containerStyle}
            className={['Field-container', inline ? 'inline' : ''].join(' ')}>
            {label
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <label htmlFor={name}>{label}</label>}
            {field}
            {error
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <span className="error">{error}</span>}
        </div>
    )
}

field.propTypes = {
  type: string,
  name: string,
  value: oneOfType([string, arrayOf(string), bool, instanceOf(Date)]),
  onChange: func,
  placeholder: string,
  error: string,
  label: string,
  list: arrayOf(string),
  rows: oneOfType([string, number]),
  disabled: bool,
  onClick: func,
  style: object,
  inline: bool,
  containerStyle: object,
  children: oneOfType([arrayOf(node), node])
}

export default field
