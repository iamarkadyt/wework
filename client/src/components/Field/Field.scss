@import 'variables.scss';
.Field-container {
    &.inline {
        display: inline;
    }
    width: 100%;
    margin: 1rem 0 0 0;
    input[type=checkbox] {
        transform: scale(1.5);
    }
    .button,
    button[type=submit] {
        min-width: 6rem;
        border-radius: .2rem;
        font-size: 1.1rem;
        padding: .2rem 1rem;
        border-color: $color-button-border;
        background: $color-button-backround;
        color: $color-button-font;
        &:active {
            border-color: $color-button-active-border;
            background: $color-button-active-background;
            color: $color-button-active-font;
        }
        &:hover:not(:active) {
            border-color: $color-button-hover-border;
            background: $color-button-hover-backround;
            color: $color-button-hover-font;
        }
    }
    button[type=submit] {
        float: right;
    }
    .button-action {
        background-color: darken(cornflowerblue, 10%);
        border-color: darken(cornflowerblue, 10%);
        background: linear-gradient(to bottom, darken(cornflowerblue, 10%), darken(cornflowerblue, 20%));
        color: white;
        &:active {
            background: linear-gradient(to bottom, darken(cornflowerblue, 25%), darken(cornflowerblue, 15%));
        }
        &:hover:not(:active) {
            background: linear-gradient(to bottom, darken(cornflowerblue, 8%), darken(cornflowerblue, 18%));
        }
    }
    .link-button {
        background: none;
        border: none;
        font-weight: bold;
        font-size: 1.1rem;
        color: $color-font-secondary;
        &:hover {
            color: $color-font-main;
            cursor: pointer;
        }
    }
    input[type=text],
    input[type=password] {
        font-size: 1rem;
        padding: .6rem;
        width: 100%;
    }
    label {
        display: inline-block;
        padding: .2rem;
    }
    textarea {
        background-color: $color-field-background;
        border-color: $color-field-border;
        color: $color-field-font;
    }
    .error {
        display: inline-block;
        color: $color-font-error;
        padding: .2rem;
    }
}