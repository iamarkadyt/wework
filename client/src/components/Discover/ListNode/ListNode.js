import React from 'react'
import { withRouter } from 'react-router-dom'
import { FaUserPlus as IcoAdd } from 'react-icons/fa'
import Field from '../../Field/Field'
import { discoverListNodeType } from '../../../types/index'
import { func, object } from 'prop-types'
import './ListNode.scss'

const ListNode = ({
    node: { _id, name, avatar, title, company },
    followAPerson,
    history
}) => (
        <div className="ListNode-container">
            <img className="ListNode-img" src={avatar} alt="" />
            <div className="ListNode-name">
                <Field
                    type="linkButton"
                    label={name}
                    inline
                    style={{ fontSize: '1.15em' }}
                    onClick={() => history.push(`/profile/id/${_id}`)} />
                <p>{title}{company && ` at ${company}`}</p>
            </div>
            <Field
                type="linkButton"
                style={{ fontSize: '1.4rem' }}
                onClick={() => followAPerson(_id)}>
                <IcoAdd />
            </Field>
        </div>
      )

ListNode.propTypes = {
  node: discoverListNodeType.isRequired,
  followAPerson: func.isRequired,
  history: object.isRequired
}

export const _UnconnectedListNode = ListNode
export default withRouter(ListNode)
