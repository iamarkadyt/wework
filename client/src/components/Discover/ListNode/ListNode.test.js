/**
 * Contract:
 *
 * What do I render? What props do I pass to components I render?
 * (props passing)
 * - ListNode-img receives avatar in src field
 * - Person name Button receives correct props
 * - paragraph right below displays title and company(if available) 
 * - [+] button receives correct props
 *
 *
 * What do I do with the props I receive? What effect do they have?
 * (props effects)
 * - if company field is present, company is shown
 * - if company is not specified, company string isn't rendered
 *
 * Interaction?
 * (interaction)
 * (args passing)
 * - Clicking on a [+] button calls followAPerson function
 * - Clicking on a person's name navigates to their profile (/profile/id/$id)
 *
 * Side effects?
 * - N/A
 *
 * State?
 * - N/A
 *
 * Context & public ref API
 * - N/A
 *
 */

import React from 'react'
import { _UnconnectedListNode as ListNode } from './ListNode'
import Field from '../../Field/Field'
import { FaUserPlus } from 'react-icons/fa'
import { shallow } from 'enzyme'
import cloneDeep from 'lodash.clonedeep'

describe('ListNode', () => {
  let props
  let mountedListNode

  const getMockProps = () => {
    return cloneDeep({
      followAPerson: jest.fn(),
      node: { 
        _id: 'bbb123',
        name: 'John Doe',
        avatar: '//gravatar.com/12345',
        title: 'Web Developer',
        company: 'Some Company'
      },
      history: {
        push: jest.fn()
      }
    })
  }

  const listNode = () => {
    if(!mountedListNode) {
      mountedListNode = shallow(<ListNode {...props} />)
    }
    return mountedListNode
  }

  beforeEach(() => {
    props = getMockProps()
    mountedListNode = undefined
  })

  describe('props passing', () => {
    it('person\'s image node receives correct props', () => {
      const avatar = 'some_avatar_123'
      const expectedProps = { 
	className: 'ListNode-img', 
	src: avatar, 
	alt: '' 
      }
      props.node.avatar = avatar
      expect(listNode().find('.ListNode-img').props()).toStrictEqual(expectedProps)
    })

    it('person\'s name button receives correct props', () => {
      // expectedProps should not depend on global props object
      const label = 'Label129'
      const expectedProps = {
        type: 'linkButton',
        label,
        inline: true,
        style: { fontSize: '1.15em' },
        onClick: expect.any(Function)
      }
      props.node.name = label
      expect(listNode().find('.ListNode-name').find(Field).props()).toStrictEqual(expectedProps)
    })

    it('paragraph below person\'s name button receives correct props', () => {
      const company = 'company_is_specified'
      const expectedProps = {
        children: [props.node.title, ` at ${company}`]
      }	
      props.node.company = company
      expect(listNode().find('.ListNode-name').find('p').props()).toStrictEqual(expectedProps)
    })

    it('+[AddPerson] button receives correct props', () => {
      const expectedProps = {
        type: 'linkButton',
        style: { fontSize: '1.4rem' },
        onClick: expect.any(Function),
        children: <FaUserPlus />
      }
      expect(listNode().find('.ListNode-container').childAt(2).props()).toStrictEqual(expectedProps)
    })
  })

  describe('args passing', () => {
    it('followAPerson function is called with correct parms', () => {
      const id = 'someID127'
      props.node._id = id
      listNode().find('.ListNode-container').childAt(2).simulate('click')
      expect(props.followAPerson).toHaveBeenCalledWith(id)
    })

    it('history.push is called with correct parms', () => {
      const id = 'some__ID_2'
      props.node._id = id
      listNode().find('.ListNode-name').find(Field).simulate('click')
      expect(props.history.push).toHaveBeenCalledWith(`/profile/id/${id}`)
    })
  })

  describe('props effects', () => {
    it('if company field is undefined, it is not rendered', () => {
      const title = 'some_title123'
      props.node.company = undefined
      props.node.title = title
      // to be purely title
      expect(listNode().find('.ListNode-name').find('p').text()).toBe(title)
    })

    it('if company field is defined, it is shown', () => {
      const company = 'some_company123'
      props.node.company = company
      expect(listNode().find('.ListNode-name').find('p').text()).toMatch(company)
    })
  })

  describe('interaction', () => {
    it('click on +[AddPerson] button calls followAPerson function', () => {
      listNode().find('.ListNode-container').childAt(2).simulate('click')
      expect(props.followAPerson).toHaveBeenCalled()
    })

    it('click on a person\'s name navigates to their profile', () => {
      listNode().find('.ListNode-name').find(Field).simulate('click')
      expect(props.history.push).toHaveBeenCalled()
    })
  })

  it('passes a snapshot test', () => {
    expect(listNode()).toMatchSnapshot()
  })
})
