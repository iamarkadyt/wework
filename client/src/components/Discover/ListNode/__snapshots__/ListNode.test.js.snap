// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListNode passes a snapshot test 1`] = `
<div
  className="ListNode-container"
>
  <img
    alt=""
    className="ListNode-img"
    src="//gravatar.com/12345"
  />
  <div
    className="ListNode-name"
  >
    <field
      inline={true}
      label="John Doe"
      onClick={[Function]}
      style={
        Object {
          "fontSize": "1.15em",
        }
      }
      type="linkButton"
    />
    <p>
      Web Developer
       at Some Company
    </p>
  </div>
  <field
    onClick={[Function]}
    style={
      Object {
        "fontSize": "1.4rem",
      }
    }
    type="linkButton"
  >
    <FaUserPlus />
  </field>
</div>
`;
